File: Backtracking_using_bitmask/Backtrack_using_bitmask.java
import java.util.*;
class Solution
{
    //Keeps track of the # of valid solutions
    static int ans = 0;

    public static int power(int x, int y)
    {
        if(y == 0)
        return 1;
        int result = power(x,y / 2);
        if(y % 2 == 0)
        return result * result;
        else
        return x * result * result;
    }
    //Checks all possible board configurations
    public static void solveNQueens(int left, int col, int right, int n, int done)
    {
        //All columns are occupied, so the solution must be complete
        if(col == done)
        {
            ans++;
            return;
        }

        //Gets a bit sequence with "1"s where ever there is an open "slot"
        int pos = ~(left | col | right) & done;

        //Loops as long as there is a valid place to put another queen.
        while(pos > 0)
        {
            int bit = pos & (-pos);
            pos = pos - bit;
            solveNQueens((left | bit) >> 1, col | bit, (right | bit) << 1, n, done);
        }
    }
    public static void main(String[] args)
    {
        Scanner input = new Scanner(System.in);
        int n = input.nextInt();
        //Helps identify valid solutions
        int done = power(2, n) - 1;
        solveNQueens(0, 0, 0, n, done);
        System.out.print(ans);
        input.close();
    }
};

/*
Sample Input:
4
Sample Output:
2
Sample Input:
5
Sample Output:
10
*/

File: Backtracking_using_bitmask/Backtrack_using_bitmask.py
<<<<<<< HEAD
<<<<<<< HEAD
#Here we use bit patterns to keep track of queen placements 
#for the columns, left diagonals and right diagonals.

def totalNQueens(n):
    #Helps identify valid solutions
    all_ones = 2 ** n - 1  
    #Keeps track of the # of valid solutions
    count = 0 
=======
# Here we use bit patterns to keep track of queen placements for the columns, left diagonals and right diagonals.

def totalNQueens(n):
    all_ones = 2 ** n - 1  #Helps identify valid solutions
    count = 0 #Keeps track of the # of valid solutions
>>>>>>> 42bb684... Backtracking using Bitmasking
=======
#Here we use bit patterns to keep track of queen placements 
#for the columns, left diagonals and right diagonals.

def totalNQueens(n):
    #Helps identify valid solutions
    all_ones = 2 ** n - 1  
    #Keeps track of the # of valid solutions
    count = 0 
>>>>>>> d83e063... Update Backtrack_using_bitmask.py

    #Checks all possible board configurations
    def helper(ld, column, rd):
        nonlocal count

        #All columns are occupied,so the solution must be complete
        if column == all_ones:
            count += 1
            return

        #Gets a bit sequence with "1"s whereever there is an open "slot"
        possible_slots = ~(ld | column | rd) & all_ones

        #Loops as long as there is a valid place to put another queen.
        while possible_slots:
            current_bit = possible_slots & -possible_slots
            possible_slots -= current_bit
            helper((ld | current_bit) >> 1,
                   column | current_bit,
                   (rd | current_bit) << 1)

    helper(0, 0, 0)
    return count

if __name__ == '__main__': 
    n = int(input())
<<<<<<< HEAD
<<<<<<< HEAD
    d = (1 << n) - 1
    cnt = totalNQueens(n)
    print(cnt);

'''
Sample Input:
4
Sample Output:
2
Sample Input:
5
Sample Output:
10
'''
=======
    d = (1<<n)-1
    cnt = totalNQueens(n)
    print(cnt);


# Sample Input: 
# 4
# Sample Output: 
# 2

# Sample Input: 
# 5
# Sample Output: 
# 10
>>>>>>> 42bb684... Backtracking using Bitmasking
=======
    d = (1 << n) - 1
    cnt = totalNQueens(n)
    print(cnt);

'''
Sample Input:
4
Sample Output:
2
Sample Input:
5
Sample Output:
10
'''
>>>>>>> d83e063... Update Backtrack_using_bitmask.py
